WIP: sway: add titlebar_v_height option

vvrein opened the pull request:
This pr is in wip state, as there is need to fix some things:

- [x] 1) We need new name for `fixed titlebar height`, since in code there already is a `titlebar_height` variable 
- [ ] 2) I need to remind myself what for `baseline` stands and find out how it can be calculated more correctly
- [x] 3) Remove or rewrite logging messages

Base logic is:
- If there is no configured `fixed titlebar height` - then variable assigns value `-1` and actual title height goes to be calculated with cairo.
- Same if there is explicit `-1` in config.
- If `fixed titlebar height` is set to 0 and above - title height is then picked from config.

----

For 1) my variants are:
- titlebar_height_fixed  <-- and my vote for this 
- titlebar_height_bind
- titlebar_height_exact

oranenj commented:
Do the padding and border width variables get used with a fixed titlebar height? I think the "v" might be redundant in the option name, so perhaps just "titlebar_height"?

Your definition of the configuration option doesn't allow "-1", but I think that's fine; if people want to make it automatic, they should just remove the configuration option. Also, allowing a height less than 1 makes no sense and probably results in weird behaviour.

You're also missing documentation for the variable, don't forget that :)

oranenj commented on sway/tree/container.c:
```diff
@@ -573,16 +573,32 @@ void container_update_title_textures(struct sway_container *container) {
 }
 
 void container_calculate_title_height(struct sway_container *container) {
+	sway_log(SWAY_INFO, "===> container_calculate_title_height");
```
I think most of this logging isn't  very useful, so to make this PR mergeable, you should remove it.

Other than that, is there anything here that's actually WIP anymore? The Alpine build failure doesn't seem related to the code.

oranenj commented on sway/tree/container.c:
```diff
@@ -573,16 +573,32 @@ void container_update_title_textures(struct sway_container *container) {
+		sway_log(SWAY_INFO, "===> config->titlebar_v_height >= 0");
+
+		height   = config->titlebar_v_height;
+		baseline = round(height * 0.77);
```
I'm not sure if I like this magic number, but I don't have any better ideas either. Any other reviewers have opinions about this?

Exagone313 replied:
Since the value is not used elsewhere, the baseline value could also be a configuration directive.

oranenj reviewed (changes requested):
A stylistic nitpick about commit messages: generally imperative mood reads better, so I personally would prefer something akin to "sway: add titlebar_v_height option to configure static titlebar height". Might even leave it as "add titlebar_v_height option" because the name is decently self-explanatory.

vvrein commented:
@oranenj Sorry me for those verbose actions... That was my auto rebuld script
<details><summary>script</summary>

```bash
#!/bin/bash
pushd sway
git fetch --all
git reset origin/master --hard
meson setup --wipe ./build

pushd subprojects/wlroots
git fetch --all
git reset origin/master --hard
popd

git rebase upstream/master
meson build --reconfigure && ninja -C build && git push origin master --force
```
</details>

I'm now going to comb the code, and take it to mergeable state. Thank for your advices :)

vvrein commented:
I found a strange thing, which can explain why I used `_v_` in parameter name:  
https://github.com/vvrein/sway/tree/titlebar_height_fixed
This code compiles, but then sway shows an error like "unknown command `titlebar_padding 0 0`"
The same is for `titlebar_height` name.

![image](https://user-images.githubusercontent.com/17022100/113485591-05132980-94b7-11eb-8965-251ad451f7db.png)


Since I could not find any obvious reason for this behavior - I better leave titlebar_v_height name for now

vvrein commented:
About magic `baseline = round(height * 0.77);`


<details><summary>screenshot</summary> 

![grimcap](https://user-images.githubusercontent.com/17022100/113486081-9d121280-94b9-11eb-894b-46d002857ef9.png)

</details>
 
 I looked at what cairo does, when it calculates baseline, and then picked up value that was near.
 ```
 >>> 12/15
0.8
>>> 12/16
0.75
>>> 14/19
0.7368421052631579
>>>
```
Also, in my experiments, baseline seems does not make any job, since hardcoding values like 10, 20, 30 does not change anything.

vvrein commented:
A little summary for now:
- I deleted all my debug logging from container_calculate_title_height
- Added a little note to manual
- Left the old variable name

Questions:
- If another than `titlebar_v_height` variable name is desirable, then I need some help with https://github.com/swaywm/sway/pull/6059#issuecomment-812893750
- Need I somehow extend note about `titlebar_v_height`?

oranenj commented:
@vvrein The `titlebar_padding` issue is because after you add titlebar_height_fixed, the array is no longer sorted. You need to add it before titlebar_padding: https://github.com/vvrein/sway/blob/b7d6ca952331261cc3c45dd51d7391a89e010fd8/sway/commands.c#L44

emersion commented:
> I looked at what cairo does, when it calculates baseline, and then picked up value that was near.

I'm not following. Can you link the cairo code? Where do these values come from?

Can't we use something like this? https://www.cairographics.org/manual/cairo-cairo-scaled-font-t.html#cairo-font-extents-t

Exagone313 commented on sway/tree/container.c:
```diff
@@ -573,16 +573,22 @@ void container_update_title_textures(struct sway_container *container) {
 }
 
 void container_calculate_title_height(struct sway_container *container) {
+        int height;
```
With project code style, these lines should be moved after the `if (!container->formatted_title)` block.

FrostKiwi commented:
Really happy with what this PR brings. I can misuse this to create the broder I always wanted: Show selection based on just a line at the top, while the title is shown in the bar.
![20210607_18h41m14s_grim](https://user-images.githubusercontent.com/60887273/121057841-fea77380-c7bf-11eb-9161-a347e603c16b.png)

vvrein commented:
This commit changes option name from `titlebar_v_height` to `titlebar_height`,   
and sets `baseline` exactly to zero, since afterf it would be updated here https://github.com/swaywm/sway/blob/master/sway/config.c#L1007-L1009

vvrein commented:
Closing in favor of https://github.com/swaywm/sway/pull/6433
